File: Data-Structure/Session-Overview/QueueDataStructure/Queue.java
package QueueDataStructure;

/**
 * @author Niraj Gupta
 *
 * Interface to implement the queue data structure and can accept any user entered data type. 
 * @param <T> : The user select the type of the queue.
 */
public interface Queue<T>{
	
	/**
	 * Method to enqueue new element to the queue.
	 * @param element : The element to be enqueueed to the queue.
	 * @return : The Queue Array in which the element is enqueueed to the front of the queue.
	 */
	Queue<T> enqueue(T element);
	
	/**
	 * Method to check if the element can be enqueueed to queue or not.
	 * @param element : The element which is to be enqueueed to the queue.
	 * @return : True if the element can be enqueueed to the queue successfully, false otherwise.
	 */
    boolean checkIfElementCanBeenqueued(T element);
    
    /**
     * Method to get the front element of the queue.
     * @return : The front element of the queue
     */
    T getFront();
    
    /**
     * Method to delete the front element from the queue
     * @return : The element which is removed from the front of the queue.
     */
    T dequeue();
    
    /**
     * Method to get the size of the queue.
     * @return : The total number of element present in the queue.
     */
    int size();
    
}

File: UI-Design-And-Development/Session-8/javascript/queue.js
/**
Queue Class or Constructor via function
*/
function Queue() {
    this.oldestIndex = 1;
    this.newestIndex = 1;
    this.storage = {};
}

/**
Method to get the size of the queue
*/
Queue.prototype.size = function () {
    return this.newestIndex - this.oldestIndex;
};

/**
Method to add or enqueue data in the queue
*/
Queue.prototype.enqueue = function (data) {
    this.storage[this.newestIndex] = data;
    this.newestIndex++;
};

/**
Method to remove data or dequeue from the queue
**/
Queue.prototype.dequeue = function () {
    var oldestIndex = this.oldestIndex,
        newestIndex = this.newestIndex,
        deletedData;
    if (oldestIndex !== newestIndex) {
        deletedData = this.storage[oldestIndex];
        delete this.storage[oldestIndex];
        this.oldestIndex++;
        return deletedData;
    }
};
